Remove duplicate public APIs where standard libbpf APIs exist

dthaler opened the pull request:
Hide as internal: ebpf_map_get_fd, ebpf_map_next, ebpf_map_previous, ebpf_object_close, ebpf_program_get_fd, ebpf_program_next, ebpf_program_previous

Remove bpf_map_delete_elem, bpf_map_delete_elem, and bpf_map_delete_elem and just use the standard names

Deal with ebpf_map name collision between the app API and 
execution context by ucontinuing to use _ebpf_map_definition in the execution context
 and aliasing it to bpf_map in ebpf_helpers.h

Signed-off-by: Dave Thaler <dthaler@microsoft.com>

Alan-Jowett commented on include/ebpf_program_types.h:
```diff
@@ -37,7 +37,7 @@ typedef struct _ebpf_helper_function_prototype
+typedef struct bpf_prog_info
 {
     ebpf_program_type_descriptor_t program_type_descriptor;
     uint32_t count_of_helpers;
```
Are the MIDL annotations still needed? I thought we stopped using RPC to serialize these structures.

shankarseal replied:
The dead code is yet to be removed. I will remove the MIDL annotations when I delete the dead code.

Alan-Jowett commented on include/ebpf_program_types.h:
```diff
@@ -37,7 +37,7 @@ typedef struct _ebpf_helper_function_prototype
 } ebpf_helper_function_prototype_t;
 
-typedef struct _ebpf_program_info
+typedef struct bpf_prog_info
```
Why bpf_prog_info and not _bpf_prog_info? I thought style guidelines had the struct name start with _?

dthaler replied:
Per our discussion this week, style guidelines are:
1) If the type is public in libbpf, match it literally,
2) else struct names start with _

In this case, rule 1 applies.

saxena-anurag replied:
Even though there is a "bpf_prog_info" in libbpf.h, it is not same as the one defined here. We had added ebpf_program_info structure to return program information from EC using the APIs `ebpf_get_program_info()` and `ebpf_get_program_info_by_fd`

dthaler replied:
thanks, reverted

Alan-Jowett commented on libs/api/api_internal.h:
```diff
@@ -23,9 +23,9 @@ typedef struct _ebpf_program
     fd_t fd;
 } ebpf_program_t;
```
Is anyone still using the ebpf_program_t typedef?

dthaler replied:
Yes there are still a bunch of uses in code in this repo

Alan-Jowett commented:
We have some name collisions in the code that we should address to avoid confusion.
typedef struct _ebpf_object
{
    char* file_name = nullptr;
    std::vector<ebpf_program_t*> programs;
    std::vector<ebpf_map_t*> maps;
} ebpf_object_t;

vs:
    typedef struct _ebpf_object
    {
        uint32_t marker;
        volatile int32_t reference_count;
        ebpf_object_type_t type;
        ebpf_free_object_t free_function;
        ebpf_list_entry_t entry;
        ebpf_list_entry_t list_entry;
    } ebpf_object_t;

I would recommend we pick a different name for the two distinct structures to avoid accidental misuse.

dthaler commented on libs/execution_context/ebpf_program.c:
```diff
@@ -11,7 +11,7 @@
 #include "ubpf.h"
 
-typedef struct _ebpf_program
+typedef struct bpf_program
```
because this file includes ebpf_link.h which has
```
    typedef struct bpf_program ebpf_program_t;
```
So if it's not renamed, the types conflict and this results in a compiler error.

dthaler replied:
I changed this and ebpf_link.h back, they just have to match each other and don't need to use a libbpf type since this is internal.

dthaler commented:
> We have some name collisions in the code that we should address to avoid confusion.
...
> I would recommend we pick a different name for the two distinct structures to avoid accidental misuse.

This comment looks orthogonal to this PR, though I dealt with a similar issue with bpf_map in this PR.

Alan-Jowett commented:
Name collision - not entirely orthogonal. It would help clarify which structures need to be renamed and which don't. Spreading libbpf names into the core execution context because we originally had a name collision seems wrong.

dthaler commented:
> Name collision - not entirely orthogonal. It would help clarify which structures need to be renamed and which don't. Spreading libbpf names into the core execution context because we originally had a name collision seems wrong.

Reverted changes that put libbpf names into the core execution context, which turned out to be necessary.   There is still a name collision with ebpf_object_t but since the execution context use is internal and separate from the user mode API one, the conflict doesn't affect anything yet.  Still I added a comment to libs/platform/ebpf_object.h warning about the collision, since that's the place that should be renamed if a rename is needed at all.

saxena-anurag commented on libs/api/ebpf_api.cpp:
```diff
@@ -67,10 +67,9 @@ ebpf_api_create_map(
-        {sizeof(struct _ebpf_map_definition), type, key_size, value_size, max_entries}};
+    _ebpf_operation_create_map_request request{sizeof(_ebpf_operation_create_map_request),
+                                               ebpf_operation_id_t::EBPF_OPERATION_CREATE_MAP,
+                                               {sizeof(struct bpf_map), type, key_size, value_size, max_entries}};
```
bpf_map is #def to _ebpf_map_definition in ebpf_helper.h which will be only included by ebpf programs. So I think we cannot use "struct bpf_map" instead of "struct _ebpf_map_definition" here?

Also bpf_map will defined as _ebpf_map in this file's context.

dthaler replied:
No that's unrelated.  The use here is a different definition of bpf_map which is defined in api_internal.h.   (There are two separate bpf_map structs in standard bpf... one in the execution context and one used in libbpf.)

saxena-anurag replied:
bpf_map in api_internal is defined as follows:

`typedef struct bpf_map
{
    const struct bpf_object* object;
    char* name;
    ebpf_handle_t map_handle;
    fd_t map_fd;
    bool pinned;
} ebpf_map_t;`

This is different than `struct _ebpf_map_definition` that we are replacing here?

dthaler replied:
reverted

saxena-anurag commented on libs/platform/ebpf_object.h:
```diff
@@ -22,6 +22,8 @@ extern "C"
     typedef void (*ebpf_free_object_t)(ebpf_object_t* object);
 
+    // This type probably ought to be renamed to avoid confusion with
+    // ebpf_object_t in libs\api\api_internal.h
```
I think Alan had once suggested that we can change the struct names in EC to ebpf_core_object, ebpf_core_program, etc.

dthaler replied:
Sure.  But left to a separate PR since not related to removing duplicate public APIs where standard public bpf apis exist

saxena-anurag commented on libs/platform/ebpf_platform.h:
```diff
@@ -707,7 +707,7 @@ extern "C"
         _In_ const ebpf_trampoline_table_t* trampoline_table, size_t index, _Out_ void** function);
 
-    typedef struct _ebpf_program_info ebpf_program_info_t;
+    typedef struct bpf_prog_info ebpf_program_info_t;
```
I think this should be _ebpf_program_info as this is in EC.

dthaler replied:
yes, changed
